dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -442,6 +442,21 @@ public Knowledge visitDefinition(Definition def) {
                     : new Knowledge(new EventGraph(), new EventGraph());
         }
 
+        @Override
+        public Knowledge visitFree(Free def) {
+            final List<Event> visibleEvents = program.getThreadEventsWithAllTags(VISIBLE);
+            EventGraph must = new EventGraph();
+            EventGraph may = new EventGraph();
+
+            for (Event e1 : visibleEvents) {
+                for (Event e2 : visibleEvents) {
+                    may.add(e1, e2);
+                }
+            }
+
+            return new Knowledge(may, enableMustSets ? must : EventGraph.empty());
```

Comment:
Must is always empty do I don't see the point of having the conditional

Reply:
I put it to have it uniform, i.e., if you decide to put a non-empty must-set later on, the option will still work. As long as we have this option, I would keep it as is. If we decide to remove the no must-set option, I would change it.

Reply:
I did not see any easy way to specify the must set of a cat-introduced relation. Thus, realistically speaking, I don't think we will ever have a non-empty must set here

Reply:
Realistically speaking we won't have non-empty must-sets. But I would like to avoid giving this one relation a different treatment that is not any more performant. I think it would just break a pattern for no benefit (with some practically unrealistic pitfalls).

Reply:
There are a few other relations that do not use the enable must set option. Even more, `visitSyncFence` explicitly uses the empty graph

Reply:
I think this is because those were made by @tonghaining and he was not aware of the existence of that option.
I think the option is implemented badly: the baseline code should simply always compute may/must-sets and the propagator engine should simply discard the must-set if the option is enabled.
Ideally, this would move the handling of that option into a single place into the code and developers don't need to think about the existence of that option when implementing their new relations/definitions.

Reply:
I have the feeling that having this fine grained option is useless. Having a way to disable the whole RA might be useful, but AFAIR Implementing this was not as trivial as it sounds. @nagacek mentioned recently that debugging certain things gets harder with so many optimizations.

Reply:
I can see if I can move that option into a central place and if not, I will just remove it.

Reply:
I moved the handling into a central place. Incidentally, I found bugs in RA when computing must-sets for Range/Domain relations.